### ISSUE TITLE
How could I append to an existing manifest in a live stream?

### ISSUE THREAD

**Issue by chrisfillmore** [google/ExoPlayer#6668]
### [REQUIRED] Searched documentation and issues
I'm new to ExoPlayer. So far I've been reading the [DASH](https://exoplayer.dev/dash.html) docs and the section on [MediaSources](https://exoplayer.dev/media-sources.html). I'm also just reading the code.

### [REQUIRED] Question
I want to know if I can build a model of a live DASH manifest which is appended to with each manifest update. This may involve literally merging periods on update, or perhaps could be achieved by some other means; I'm still working on the details. Basically what we want to do:
- User starts watching a stream at the live edge
- After watching for an hour, the manifest now has an hour's worth of content listed, and user could e.g. seek back
- Or similar use cases, e.g. user tunes in and pauses for an hour, then is able to continue playback from the paused point, and seek forward if they want

I'm not sure where in ExoPlayer's lifecycle hooks / dependencies, this kind of feature could be implemented. I would be grateful for any advice. Please let me know if you have questions.

### A full bug report captured from the device
N/A

### Link to test content
N/A

**Comment by ojw28** [comment]
If you append to a DASH manifest then yes, ExoPlayer will do the right thing. How to update the DASH manifest isn't an ExoPlayer specific question.

> This may involve literally merging periods on update

It's unclear what this means, but it doesn't sound like a valid way of updating a DASH manifest.

**Comment by chrisfillmore** [comment]
Hi @ojw28 thanks for replying. I think I may have been unclear. Let me try again:

Given that I tune into the live edge of a stream with `timeShiftBufferDepth="PT3M"` and watch for 15 minutes, when I attempt to seek back 15 minutes in time, then ExoPlayer will seek to that point and playback will continue.

In general I know that this process can work. My team has an internal DASH player which supports this behaviour, but I want to know if I can achieve the same thing in ExoPlayer. In our use case, we know that the segments are still available beyond the initial stated `timeShiftBufferDepth`.

I expect it would work something like:
- On every manifest update, update `timeShiftBufferDepth` to be `timeShiftBufferDepth` + time elapsed since stream started
- Find the union of segments across periods

Thus when the player goes to seek / resume playback, it has access to all segments available since the user started watching a stream.

Is that clearer?

**Comment by chrisfillmore** [comment]
My latest thought is that I could do this using HttpDataSource, by intercepting manifest responses. By maintaining my own model of manifest state, and updating it with each response, and then returning the updated manifest to ExoPlayer, then the player would allow the behaviour I want.

Do you suppose this would be the best way?

**Comment by chrisfillmore** [comment]
Newer idea...
When I call `new DashMediaSource.Factory(...)` I provide a DataSource.Factory (as manifestDataSourceFactory). It seems I could provide a custom implementation here which would handle the "merging" of manifests that I want to do.

**Comment by ojw28** [comment]
So, you have a DASH live stream that has a fixed size window of 3 minutes, and what you want is for the client to treat it as a DASH live stream whose window grows over time. Why don't you just serve the DASH manifest that you want in the first place, rather than serving a different one and then relying on client side hacks to reconstruct what you actually wanted?

**Comment by chrisfillmore** [comment]
> Why don't you just serve the DASH manifest that you want in the first place, rather than serving a different one and then relying on client side hacks to reconstruct what you actually wanted?

Yeah you're preaching to the choir on this point. This is not something I can control. My team is pursuing this option because it is the _least bad_ hack available to us. And, unfortunately, I can't go into a lot of detail about our implementation. Please forgive me for any vagueness as a result. But as you can likely infer, there are multiple services available to a client for a given stream, and the client has business logic to decide when/how to use them.

For the sake of argument: as justification, I would say that in any large organization trying to deliver a streaming video product, groups in that organization will have goals to deliver features within timeframes that constrain those groups' abilities to do it "the right way". We may eventually arrive at "the right way" (say, 6 months or a year from now), but we have to deliver a feature sooner than that. Hence, hacks. In my observation, this is the difference between working on a consumer software application, and a turnkey middleware library like ExoPlayer.

Back to the point... should I do this by injecting a custom DataSource.Factory as manifestDataSourceFactory? This sounds like the most direct approach.

**Comment by ojw28** [comment]
The problem with doing it at the `DataSource` level is that those components are dealing with bytes rather than structured data. So you'd have to deserialize the manifest, merge it and reserialize it again.

A more direct route is probably to use `DashMediaSource.Factory.setManifestParser` to pass a custom manifest parser, which would look something like this:

```
public final class CustomDashManifestParser implements ParsingLoadable.Parser<DashManifest> {

  private final DashManifestParser defaultParser = new DashManifestParser();
    
  private DashManifest mergedManifest;
    
  @Override
  public DashManifest parse(Uri uri, InputStream inputStream) throws IOException {
    DashManifest newManifest = defaultParser.parse(uri, inputStream);
    mergedManifest = mergeManifests(mergedManifest, newManifest);
    return mergedManifest;
  }
    
  private DashManifest mergeManifests(
      DashManifest previousManifest, DashManifest newManifest) {
    // TODO: Implement this
  }
}
```

You'd have to be careful not to share `CustomDashManifestParser` instances across multiple `DashMediaSource`s, since your custom parser will be stateful.

**Comment by chrisfillmore** [comment]
Thank you for the tip! I will look at that.

I'll close for now.
